File: 4119-minimum-distance-between-three-equal-elements-ii/minimum-distance-between-three-equal-elements-ii.cpp
class Solution {
public:
    int minimumDistance(vector<int>& nums) {
        int n = nums.size();
        unordered_map<int, vector<int>> mp;
        
        for (int i = 0; i < n; i++) {
            mp[nums[i]].push_back(i);
        }

        int ans = INT_MAX;

        for (auto &p : mp) {
            vector<int> &a = p.second;
            int s = a.size();
            if (s < 3) continue;
            for (int i = 0; i + 2 < s; ++i) {
                ans = min(ans, 2 * (a[i+2] - a[i]));
            }


           
        }

        return ans == INT_MAX ? -1 : ans;
    }
};

File: 4119-minimum-distance-between-three-equal-elements-ii/minimum-distance-between-three-equal-elements-ii.java
class Solution {
    public int minimumDistance(int[] nums) {
        int n = nums.length;
        int ans = (int)1e9;
        Map<Integer,List<Integer>>map = new HashMap<>();
        for(int i = 0; i<n; i++){
            if (!map.containsKey(nums[i])) {
                  map.put(nums[i], new ArrayList<>());
            }
            map.get(nums[i]).add(i);
            List<Integer> list = map.get(nums[i]);
            int m = list.size();
            if(m < 3) continue;

            int k = list.get(m-1);
            int j = list.get(m-3);

            ans = Math.min(ans, 2*(k-j));

            
        }
return ans == (int)1e9 ? -1 : ans;
        
        
    }
}